stan/math/rev/mat/functor/cvodes_integrator.hpp

```diff
@@ -0,0 +1,383 @@
+
+      void check_flag(int flag, const std::string& func_name) const {
+        if (flag < 0) {
+          std::ostringstream ss;
+          ss << func_name << " failed with error flag " << flag;
+          throw std::runtime_error(ss.str());
+        }
+      }
+
+      void set_cvode_options(double rel_tol,
+                             double abs_tol,
+                             long int max_num_steps  // NOLINT(runtime/int)
+                             ) {
+        // Forward CVode errors to noop error handler
+        CVodeSetErrHandlerFn(cvode_mem_, silent_err_handler_cvodes, 0);
+
+        check_flag(CVodeInit(cvode_mem_, &ode::ode_rhs, t0_, cvode_state_),
+                    "CVodeInit");
+
+        // Assign pointer to this as user data
+        check_flag(CVodeSetUserData(cvode_mem_,
+                                     reinterpret_cast<void*>(this)),
+                    "CVodeSetUserData");
+
+        // Initialize solver parameters
+        check_flag(CVodeSStolerances(cvode_mem_, rel_tol, abs_tol),
+                    "CVodeSStolerances");
+
+        check_flag(CVodeSetMaxNumSteps(cvode_mem_, max_num_steps),
+                    "CVodeSetMaxNumSteps");
+
+        double init_step = 0;
+        check_flag(CVodeSetInitStep(cvode_mem_, init_step),
+                    "CVodeSetInitStep");
+
+        long int max_err_test_fails = 20;  // NOLINT(runtime/int)
+        check_flag(CVodeSetMaxErrTestFails(cvode_mem_, max_err_test_fails),
+                    "CVodeSetMaxErrTestFails");
+
+        long int max_conv_fails = 50;  // NOLINT(runtime/int)
+        check_flag(CVodeSetMaxConvFails(cvode_mem_, max_conv_fails),
+                    "CVodeSetMaxConvFails");
+      }
+
+    public:
+      typedef boost::is_same<T1, stan::math::var> initial_var;
+      typedef boost::is_same<T2, stan::math::var> param_var;
+      typedef std::vector<std::vector<
+                            typename stan::return_type<T1, T2>::type>
+                          > return_t;
+
+      /**
+       * Construct CVODES integrator for an ODE model with initial
+       * state, initial time, flag for sensitivity calculation of the
+       * initial values and for the parameters, integrator options
+       * such as rel+abs tolerance and maximum number of steps. The
+       * integrator supports non-stiff (Adams-Moulton) and stiff (BDF)
+       * integration with optional stability detection.
+       *
+       * The integrator creates as output a vector of vector
+       * format. The outer vector is over the time-points where the
+       * solution is requested and the inner the states. The order of
+       * the states with is always
+       *
+       * \f[
+       * (y, \frac{\partial y}{\partial y_0}, \frac{\partial y}{\partial \theta}).
+       * \f]
+       *
+       * While the first N states correspond to y, all the remaining
+       * are optional and depend on flags given to the constructor.
+       *
+       * @param[in] ode_model functor.
```

Comment:
Doxygen is properly flagging this as a problem. There's no `ode_model` argument to the `cvodes_constructor`.
`ode_model` needs to be replaced with `f`. Every `@param` should line up with every parameter. 

This is just nitpicking and you don't need to do this. Since this is a constructor and all things are inputs, I think we should just remove all the `[in]`. I really hope none of our code is modifying inputs through constructors....